README
Minimal Python binding for Gnu FriBidi.

It used to work with FriBidi on 2006.


ChangeLog
pyfribidi 0.5.0

  API Changes:

 * Nir Soffer: Accept either unicode or encoded string or subclasses
 * Nir Soffer: Accept encoding keyword argument, support any encoding Python
   support (encoding it ignored if you send unicode input)
 * Nir Soffer: Accept base_direction keyword argument (was only positional before)
 * Nir Soffer: Return the same type of input you sent, either unicode or str
 * Nir Soffer: Updated docstrings

  Internal Changes:

 * Nir Soffer: new function log2vis_encoded_string handles strings using either
   log2vis_utf8 or log2vis_unicode
 * Nir Soffer: Updated docs
 * Nir Soffer: Added more tests, organized tests by types

 -- kobi <kzamir@walla.co.il>  Thu, 18 Jan 2006 00:59:24 +0200

pyfribidi 0.3.6

  * Nir Soffer: Raise ValueError for invalid base_dir_string.
  * Nir Soffer: Added detailed docstring
  * Nir Soffer: Use real unittests, add some tests.

 -- kobi <kzamir@walla.co.il>  Thu, 12 Jan 2006 00:59:24 +0200

pyfribidi 0.3.5

  * Nir Soffer: add suggestion to use setup.py in the INSTALL file.
  * add fribidi-systray.py example, this example use GTK and creates a system tray
    icon that convert the content of the clipboard.

 -- kobi <kzamir@walla.co.il>  Thu, 12 Jan 2006 00:59:24 +0200

pyfribidi 0.3.2

  * fix configure script to check for python devel files
  * add base_dir opetion to the log2vis python function

 -- kobi <kzamir@walla.co.il>  Fri, 18 Nov 2005 00:59:24 +0200

pyfribidi 0.3.1

  * adopted package for ubuntu
  * build python2.3 and python2.4 packages now
  * use cdbs instead of plain debhelper

 -- Stefan Potyra <sistpoty@ubuntu.com>  Wed, 16 Nov 2005 22:30:36 +0100

pyfribidi 0.3.0

  * FIXME: Allways use fribidi Right-To-Left embedding, for stas

 -- kobi <kzamir@walla.co.il>  Thu, 10 Nov 2005 08:59:24 +0200
 
pyfribidi 0.2.0

  * Tzafrir Cohen: vis2log -> log2vis

 -- kobi <kzamir@walla.co.il>  Thu, 10 Nov 2005 08:59:24 +0200

pyfribidi 0.1.0-1

  * Initial release

 -- kobi <kzamir@walla.co.il>  Thu, 10 Nov 2005 00:59:24 +0200


test_pyfribidi.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
"""
These are very basic tests, because fribidi has its own tests. It may be
better to reuse fribidi own tests, but its not clear what is the value
of base_dir for those tests.
"""

import syspath
import unittest
import pyfribidi
from pyfribidi import RTL, LTR, ON


class TestSkipped(Exception):
    """ Raised when test can not run """


class InputTests(unittest.TestCase):

    def testRequireInputString(self):
        """ input: require input string or unicode """
        self.assertRaises(TypeError, pyfribidi.log2vis)
        
    def testInvalidInputString(self):
        """ input: raise TypeError for non string or unicode input """
        self.assertRaises(TypeError, pyfribidi.log2vis, 1)

    def testInvalidDirection(self):
        """ input: raise ValueError for invalid direction """
        self.assertRaises(ValueError, pyfribidi.log2vis, u"שלום",
                          base_direction=1)

    def testUnknownEncoding(self):
        """ input: raise LookupError for invalid encoding """
        self.assertRaises(LookupError, pyfribidi.log2vis, "שלום",
                          encoding='foo')

    def testInvalidEncodedString(self):
        """ input: raise UnicodeError for invalid encoded string """
        self.assertRaises(UnicodeError, pyfribidi.log2vis, "שלום",
                          encoding='iso8859-8')
        
        
class UnicodeTests(unittest.TestCase):

    def testEmpty(self):
        """ unicode: empty string """
        self.assertEqual(pyfribidi.log2vis(u''), u'')

    def testBigString(self):
        """ unicode: big string

        It does not make sense to order such big strings, this just
        checks that there are no size limits in pyfribidi.
        """
        # About 2MB string for default python build (ucs2)
        big = (u'א' * 1024) * 1024
        self.assertEqual(pyfribidi.log2vis(big), big)

    def disabled_testMemoryFailure(self):
        """ unicode: raise MemoryError when memory allocation fails

        Disabled because it may not work on your machine. If you have
        very little memory, it may fail before calling pyfribidi, if you
        have too much memory, it may fail because no MemoryError was
        raised.
        """
        # About 512MB string for default python build (ucs2)
        try:
            huge = ((u'x' * 1024) * 1024) * 256
        except MemoryError:
            raise TestSkipped('not enough memory to create test string')
        
        self.assertRaises(MemoryError, pyfribidi.log2vis, huge)

    def testDefaultDirection(self):
        """ unicode: use RTL default """
        self.assertEqual(pyfribidi.log2vis(u"hello - שלום"),
                         pyfribidi.log2vis(u"hello - שלום", RTL))

    def testAsRTL(self):
        """ unicode: reorder line as RTL """
        self.assertEqual(pyfribidi.log2vis(u"hello - שלום", RTL),
                         u"םולש - hello")

    def testAsLTR(self):
        """ unicode: reorder line as LTR """
        self.assertEqual(pyfribidi.log2vis(u"hello - שלום", LTR),
                         u"hello - םולש")

    def testNaturalLTR(self):
        """ unicode: reorder LTR line by natural order """
        self.assertEqual(pyfribidi.log2vis(u"hello - שלום", ON),
                         u"hello - םולש")

    def testNaturalRTL(self):
        """ unicode: reorder RTL line by natural order """
        self.assertEqual(pyfribidi.log2vis(u"שלום - hello", ON),
                         u"hello - םולש")
                             
    
class UTF8Tests(unittest.TestCase):
    """ Same tests for utf8, used mainly on linux """
    
    def testEmpty(self):
        """ utf8: empty string """
        self.assertEqual(pyfribidi.log2vis(''), '')

    def testBigString(self):
        """ utf8: big string

        It does not make sense to order such big strings, this just
        checks that there are no size limits in pyfribidi.
        """
        # About 2MB string
        big = ('א' * 1024) * 1024
        self.assertEqual(pyfribidi.log2vis(big), big)

    def testDefaultDirection(self):
        """ utf8: use RTL default """
        self.assertEqual(pyfribidi.log2vis("hello - שלום"),
                         pyfribidi.log2vis("hello - שלום", RTL))
    
    def testAsRTL(self):
        """ utf8: reorder line as RTL """
        self.assertEqual(pyfribidi.log2vis("hello - שלום", RTL),
                         "םולש - hello")

    def testAsLTR(self):
        """ utf8: reorder line as LTR """
        self.assertEqual(pyfribidi.log2vis("hello - שלום", LTR),
                         "hello - םולש")
    
    def testNaturalLTR(self):
        """ utf8: reorder LTR line by natural order """
        self.assertEqual(pyfribidi.log2vis("hello - שלום", ON),
                         "hello - םולש")
    
    def testNaturalRTL(self):
        """ utf8: reorder RTL line by natural order """
        self.assertEqual(pyfribidi.log2vis("שלום - hello", ON),
                         "hello - םולש")


class OtherEncodingsTests(unittest.TestCase):
    """ Minimal tests for other encodings """
    
    def testIso8859_8NaturalRTL(self):
        """ other encodings: iso8859-8 """
        charset = 'iso8859-8'
        self.assertEqual(pyfribidi.log2vis(u"שלום - hello".encode(charset),
                                           encoding=charset),
                         u"hello - םולש".encode(charset))
    
    def testCp1255NaturalRTL(self):
        """ other encodings: cp1255 """
        charset = 'cp1255'
        self.assertEqual(pyfribidi.log2vis(u"שלום - hello".encode(charset),
                                           encoding=charset),
                         u"hello - םולש".encode(charset))

    def testUTF16NaturalRTL(self):
        """ other encodings: utf-16 """
        charset = 'utf-16'
        self.assertEqual(pyfribidi.log2vis(u"שלום - hello".encode(charset),
                                           encoding=charset),
                         u"hello - םולש".encode(charset))

        
if __name__ == '__main__':
    suite = unittest.defaultTestLoader.loadTestsFromName(__name__)
    unittest.TextTestRunner(verbosity=2).run(suite)



pyfribidi.h

/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

/* Copyright (C) 2005,2006 Yaacov Zamir, Nir Soffer */

/* 	FriBidi python binding:
    
    Install:
	python setup.py install
    
 */

#ifndef __PYFRIBIDI_H__
#define __PYFRIBIDI_H__

static PyObject *_pyfribidi_log2vis (PyObject * self, PyObject * args,
				     PyObject * kw);

static PyObject *log2vis_unicode (PyObject * unicode,
				  FriBidiCharType base_direction);

static PyObject *log2vis_encoded_string (PyObject * string,
					 const char *encoding,
					 FriBidiCharType base_direction);

static PyObject *log2vis_utf8 (PyObject * string, int unicode_length,
			       FriBidiCharType base_direction);

void initpyfribidi (void);

PyDoc_STRVAR (_pyfribidi__doc__,
	      "simple Python binding for fribidi.\n\n"
	      "pyfribidi uses libfribidi to order text visually using the unicode\n"
	      "algorithm. pyfribidi can also convert text from visual order to\n"
	      "logical order, but the conversion may be wrong in certain cases.\n");

/* Based on fribidi-0.10.7 README */
PyDoc_STRVAR (_pyfribidi_log2vis__doc__,
	      "log2vis(logical, base_direction=RTL , encoding='utf-8') -> visual\n\n"
	      "Return string reordered visually according to base direction.\n"
	      "Return the same type of input string, either unicode or string using\n"
	      "encoding.\n\n"
	      "Note that this function does not handle line breaking. You should\n"
	      "call log2vis with each line.\n\n"
	      "Arguments:\n\n"
	      "- logical: unicode or encoded string\n"
	      "- base_direction: optional logical base direction. Accepts one of\n"
	      "  the constants LTR, RTL or ON, defined in this module. ON calculate\n"
	      "  the base direction according to the BiDi algorithm.\n"
	      "- encoding: optional string encoding (ignored for unicode input)\n");

#endif /* __PYFRIBIDI_H__ */


fribidi_systray.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

# Copyright (C) 2005 Yaacov Zamir

import gtk
import egg.trayicon
import pyfribidi

clip = gtk.Clipboard()
button = gtk.Button("BD")
trayIcon = egg.trayicon.TrayIcon("BiDiTrayIcon")

def on_button_clicked(obj):
	inText = clip.wait_for_text()
	outText = pyfribidi.log2vis (inText, pyfribidi.ON)
	clip.set_text(outText)

button.connect("clicked", on_button_clicked)
trayIcon.add(button)
trayIcon.show_all()

gtk.main()


pyfribidi.c

/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

/* Copyright (C) 2005,2006 Yaacov Zamir, Nir Soffer */

/* 	FriBidi python binding:
    
    Install:
	python setup.py install
    
 */

#include <Python.h>		/* must be first */
#include <fribidi.h>
#undef _POSIX_C_SOURCE

#include "pyfribidi.h"

static PyObject *
_pyfribidi_log2vis (PyObject * self, PyObject * args, PyObject * kw)
{
	PyObject *logical = NULL;	/* input unicode or string object */
	FriBidiCharType base = FRIBIDI_TYPE_RTL;	/* optional direction */
	const char *encoding = "utf-8";	/* optional input string encoding */

	static char *kwargs[] =
		{ "logical", "base_direction", "encoding", NULL };

	if (!PyArg_ParseTupleAndKeywords (args, kw, "O|ls", kwargs,
					  &logical, &base, &encoding))
		return NULL;

	/* Validate base */

	if (!(base == FRIBIDI_TYPE_RTL ||
	      base == FRIBIDI_TYPE_LTR || base == FRIBIDI_TYPE_ON))
		return PyErr_Format (PyExc_ValueError,
				     "invalid value %d: use either RTL, LTR or ON",
				     base);

	/* Check object type and delegate to one of the log2vis functions */

	if (PyUnicode_Check (logical))
		return log2vis_unicode (logical, base);
	else if (PyString_Check (logical))
		return log2vis_encoded_string (logical, encoding, base);
	else
		return PyErr_Format (PyExc_TypeError,
				     "expected unicode or str, not %s",
				     logical->ob_type->tp_name);
}

/*
  log2vis_unicode - reorder unicode string visually

  Return value: new reference

  Return Python unicode object ordered visually or NULL if an exception
  was raised.

  Since Python and fribidi don't now know each other unicode format,
  encode input string as utf-8 and invoke log2vis_utf8.
  
  Arguments:
  
  - unicode: Python unicode object
  - base_direction: input string base direction, e.g right to left  
*/

static PyObject *
log2vis_unicode (PyObject * unicode, FriBidiCharType base_direction)
{
	PyObject *logical = NULL;	/* input string encoded in utf-8 */
	PyObject *visual = NULL;	/* output string encoded in utf-8 */
	PyObject *result = NULL;	/* unicode output string */

	int length = PyUnicode_GET_SIZE (unicode);

	logical = PyUnicode_AsUTF8String (unicode);
	if (logical == NULL)
		goto cleanup;

	visual = log2vis_utf8 (logical, length, base_direction);
	if (visual == NULL)
		goto cleanup;

	result = PyUnicode_DecodeUTF8 (PyString_AS_STRING (visual),
				       PyString_GET_SIZE (visual), "strict");

      cleanup:
	Py_XDECREF (logical);
	Py_XDECREF (visual);

	return result;
}

/*
  log2vis_encoded_string - reorder encoded string visually

  Return value: new reference

  Return Python string object ordered visually or NULL if an exception
  was raised. The returned string use the same encoding.

  Invoke either log2vis_utf8 or log2vis_unicode.

  - string: Python string object using encoding
  - encoding: string encoding, any encoding name known to Python
  - base_direction: input string base direction, e.g right to left
 */

static PyObject *
log2vis_encoded_string (PyObject * string, const char *encoding,
			FriBidiCharType base_direction)
{
	PyObject *logical = NULL;	/* logical unicode object */
	PyObject *result = NULL;	/* output string object */

	/* Always needed for the string length */
	logical = PyUnicode_Decode (PyString_AS_STRING (string),
				    PyString_GET_SIZE (string),
				    encoding, "strict");
	if (logical == NULL)
		return NULL;

	if (strcmp (encoding, "utf-8") == 0)
		/* Shortcut for utf8 strings (little faster) */
		result = log2vis_utf8 (string,
				       PyUnicode_GET_SIZE (logical),
				       base_direction);
	else
	{
		/* Invoke log2vis_unicode and encode back to encoding */

		PyObject *visual = log2vis_unicode (logical, base_direction);

		if (visual)
		{
			result = PyUnicode_Encode (PyUnicode_AS_UNICODE
						   (visual),
						   PyUnicode_GET_SIZE (visual),
						   encoding, "strict");
			Py_DECREF (visual);
		}
	}

	Py_DECREF (logical);

	return result;
}

/*
  log2vis_utf8 - reorder string visually

  Return value: new reference

  Return Python string object ordered visually or NULL if an exception
  was raised.
  
  Arguments:
  
  - string: Python string object using utf-8 encoding
  - unicode_length: number of characters in string. This is not the
    number of bytes in the string, which may be much bigger than the
    number of characters, because utf-8 uses 1-4 bytes per character.
  - base_direction: input string base direction, e.g right to left
*/

static PyObject *
log2vis_utf8 (PyObject * string, int unicode_length,
	      FriBidiCharType base_direction)
{
	FriBidiChar *logical = NULL;	/* input fribidi unicode buffer */
	FriBidiChar *visual = NULL;	/* output fribidi unicode buffer */
	PyObject *result = NULL;	/* failure */

	/* Allocate fribidi unicode buffers */

	logical = PyMem_New (FriBidiChar, unicode_length + 1);
	if (logical == NULL)
	{
		PyErr_SetString (PyExc_MemoryError,
				 "failed to allocate unicode buffer");
		goto cleanup;
	}

	visual = PyMem_New (FriBidiChar, unicode_length + 1);
	if (visual == NULL)
	{
		PyErr_SetString (PyExc_MemoryError,
				 "failed to allocate unicode buffer");
		goto cleanup;
	}

	/* Convert to unicode and order visually */

	fribidi_utf8_to_unicode (PyString_AS_STRING (string),
				 PyString_GET_SIZE (string), logical);

	if (!fribidi_log2vis (logical, unicode_length, &base_direction, visual,
			      NULL, NULL, NULL))
	{
		PyErr_SetString (PyExc_RuntimeError,
				 "fribidi failed to order string");
		goto cleanup;
	}

	/* Create result string and encode the reordered string into it */

	result = PyString_FromStringAndSize (NULL, PyString_GET_SIZE (string));
	if (result == NULL)
		/* XXX does it raise any error? */
		goto cleanup;

	fribidi_unicode_to_utf8 (visual, unicode_length,
				 PyString_AS_STRING (result));

      cleanup:
	/* Delete unicode buffers */
	PyMem_Del (logical);
	PyMem_Del (visual);

	return result;
}

static PyMethodDef PyfribidiMethods[] = {
	{"log2vis", (PyCFunction) _pyfribidi_log2vis,
	 METH_VARARGS | METH_KEYWORDS,
	 _pyfribidi_log2vis__doc__},
	{NULL, NULL, 0, NULL}
};

PyMODINIT_FUNC
initpyfribidi (void)
{
	PyObject *module;

	/* XXX What should be done if we fail here? */

	module = Py_InitModule3 ("pyfribidi", PyfribidiMethods,
				 _pyfribidi__doc__);

	PyModule_AddIntConstant (module, "RTL", (long) FRIBIDI_TYPE_RTL);
	PyModule_AddIntConstant (module, "LTR", (long) FRIBIDI_TYPE_LTR);
	PyModule_AddIntConstant (module, "ON", (long) FRIBIDI_TYPE_ON);

	PyModule_AddStringConstant (module, "__author__",
				    "Yaacov Zamir and Nir Soffer");
}
